### ISSUE TITLE
Sorting CORSIKA showers

### ISSUE THREAD

**Issue by 153957** [HiSPARC/sapphire#72]
With the merging of #70 I started a cronjob that creates jobs on Stoomboot to sort the simulations.
Im not certain if it is very smart to run so many jobs at the same time, it may be IO limited, not CPU limited. However, the storage server should be quite fast (40 TB RAID 6, I think)

The jobs are currently run on the `short` queue of Stoomboot, here jobs have a limit of 4 hours of walltime. I started the jobs about 3-4 hours ago. some of the larger showers (5 GB `corsika.h5` file) are still running (been running for 3 hours). The output (`corsika_sorted.h5`) is already 4 GB, so the job should be near completion. This does mean that showers larger than 5 GB might not make it within the 4 hours.

Those jobs that do not make it should fail and not create a `tmp_sorted_flag` (a file to mark the shower as sorted), because of this they are not removed from the `sort_queued.log`, so no new job will be started for them. This is good. We can wait (several days) until an attempt has been made to convert all showers, then wait until all jobs finish. Whatever remains in `sort_queued.log` will need to be resubmitted on the `generic` queue (change submit command in the script)

**Comment by davidfokkema** [comment]
👍

**Comment by 153957** [comment]
For instance one that fails:
`564284027_194135449`
At wall time 3:55 this is the status:
```
7.6G corsika.h5
3.6G corsika_sorted.h5
```

**Comment by 153957** [comment]
I've switched new jobs to the generic queue.
Some jobs will need to be rerun.

**Comment by 153957** [comment]
First pass is done, some showers need to be rerun because they were on short queue but should have used the generic.
(Also still some showers left to be properly converted to HDF5.)

**Comment by 153957** [comment]
Sorting already converted showers was done, however, some showers were not yet properly stored as HDF5. An attempt to store them as HDF5 and sort them failed because of low disk space. Some cleaning up is required. All temporary PyTables files have been removed, but some unsorted corsika.h5 and possibly damage files may be left.

Unfortunately this also affected the raw datastorage, since that is stored on the same server/partition.
Most stations were unable to upload data this weekend, however, removing the tmp pytables files cleared enough space to resume datastorage. All stations have already offloaded their locally buffered events.

We should look into removing some of the unneeded DAT000000 files, since those that have been converted to h5 files successfully should™ no longer be required, starting with the small showers which can be regenerated with relative ease..

**Comment by davidfokkema** [comment]
It is important to watch disk space, then. Glad that we have a fail-safe system in place for data storage!

**Comment by 153957** [comment]
I've cleaned up the temporary HDF5 files left after the issues mentioned above.
I believe the issue did not start because of low disk space. I tried rerunning the conversion of a single large shower, with currently plenty of free space. However, that failed with a segfault. I thought this might be due to insufficient RAM, possibly while creating the CSIndex. The 'storing sorted' part should be safe from insufficient RAM (says so in the documentation).

I did come across this commit for the new v3.2.1 release: https://github.com/PyTables/PyTables/commit/c252aeef76e8a19753fa3a4772ec9d099388f3c3
It seems to fix some Seg fault, perhaps the one troubling us. So, update PyTables to 3.2.1 and try again.

**Comment by 153957** [comment]
Possibly related: https://github.com/PyTables/PyTables/commit/d97dd1e070c110e608a758fb0745023aff740a23
Will be fixed in v3.2.2

**Comment by 153957** [comment]
v3.2.2 is released, but not available via conda yet.

**Comment by 153957** [comment]
PyTables v3.2.2 is not also available via conda.
I started a new store job for `639284940_464713283` (1e18 eV, proton, 30°, 14GB) on Stoomboot.

**Comment by 153957** [comment]
The job completed on Oct 2 at 5:03. So it took almost 40 hours... the resulting  HDF5 file is 19GB.
To properly support the even larger we may have to increase the job walltime to 96 hrs to be safe.

**Comment by 153957** [comment]
Several large (10^18 eV, < 30°) showers failed to be stored and sorted within the 96 hours.
Some did manage to be stored in the unsorted maner. So some cases would benefit from simply splitting the jobs into two parts, first convert, then in a new job sort.

For now Im performing the sorting of several of those big showers locally.

**Comment by 153957** [comment]
Well, that failed to, after several days of mostly reading it crashed..

**Comment by davidfokkema** [comment]
Wow! What seems to be the problem? A very large number of particles compared to the other showers?

**Comment by 153957** [comment]
Yes, it is a very large number, the HDF5 file for this particular shower is about 40 GB.
The strange this is that it also includes the CSIndex that is required by PyTables to copy a sorted version of a table. So all it really needs to do is read the index and simply go through it, reading the parts from the file that contain the particles it wants and write them to the output.

Here is the Disk Activity of the drive from which the data was read, and was also being written to:
![disk_activity](https://cloud.githubusercontent.com/assets/1433682/11285610/b2ea11e8-8f10-11e5-863b-f822e6054d2f.png)
The 'attempt' lasted about 4 days, almost continuously reading at 120 MB/s.
So a total of about 4 x 24 x 60 x 60 seconds x 120 MB/s = 350 000 seconds x 120 MB/s = 42 TB...
But it had written only a couple GB.. something seems very broken here...

**Comment by 153957** [comment]
Reported here: PyTables/PyTables#516

**Comment by davidfokkema** [comment]
Seems to be a PyTables issue, indeed. Hopefully it will be easy to track down!

**Comment by tomkooij** [comment]
I was curious, so I investigated a bit:

42 TB reading per couple GB writing might be correct. Sorting row by row means "read an entire chunk, select a few rows, read next chunk". The file is thus read thousands of times. For >40Gb this is very painful.

pseudo code of :https://github.com/PyTables/PyTables/blob/develop/tables/table.py#L2778-2788
```
foreach bunch of rows in sorted_table:
     use index to calc coordinates 
     read_rows(coordinates)
```
For each call to read_rows **every chunk that is "hit" needs to be read from disk**. 

For a 30Gb testfile read_rows is called 30.000 times... Each read_row can easily read 10% of the file from disk.

If sorting huge showers is useful, It should be possible to implement an algorithm that scales better.

**Comment by 153957** [comment]
Perhaps it would help to reduce the chunksize of the 'source' file?

**Comment by tomkooij** [comment]
We can try a few chunksizes to see if that helps. For a small testfile we need to test 

If we really want better sorts a different (non-

**Comment by davidfokkema** [comment]
I haven't had a chance to look at the code, but is it possible to read in the columns which need to be sorted, perform the sorting in memory to figure out the sorting order and then read the original file out of order and write the new file in the correct order? Maybe that's exactly what the code does.

**Comment by tomkooij** [comment]
@davidfokkema: That is what the code does (using an index).

I tried:
```python
import tables
from sapphire.utils import pbar

FILENAME = 'corsika_unsorted.h5'

with tables.open_file(FILENAME,'r') as data:
    t = data.get_node('/groundparticles')

    nrows = len(t)
    print "nrows: ", nrows

    old, new = (-1e99,-1e99)
    for row in pbar(t.itersorted('x'), length=nrows):
        old, new = new, row['x']
        assert new >= old
```
Which basically just reads the file in order using an index. This takes >20hours for a 30Gb file, while the full sort is about 27 hours (both expected, not actual). It is possible that setting ```chunksize``` smaller helps, but I doubt it will really help.

This algorithm is O(n^2). There are probably O(n*log(n)) sorting algorthms which are optimised for "on-disk". But I wonder if it is worth it.

Option 1: Use the code above and just ```dest_row.append()``` each row and hope it does not crash).
Option 2: Google an on-disk sort algorithm that is better than O(n^2) and implement.

**Comment by 153957** [comment]
@davidfokkema Yes for the most part :)

Currently we do the following:
- First convert the DAT000000 file to an HDF5 file. Reading out the particles one by one from the file, quite randomly sorted.
- Then for the HDF5 file a CSIndex (completely sorted index) is created for the x-column.
- This will be used by PyTables when using the copy method with `sortby='x'` to find the order in which the rows need to be copied.

**Comment by 153957** [comment]
Chunksize might make a difference if the data is on an SSD, so small random reads are fast..

**Comment by davidfokkema** [comment]
So, if we were to just read all rows in a random order, without using any index or sorting, it is still very, very slow?

**Comment by tomkooij** [comment]
No thats very fast: Reading the 30Gb file with ```table.read()``` (in chunks) is just 13 seconds.

The solution seems to be on-disk mergesort. Very easy to implement. But I need to do other stuff...

**Comment by davidfokkema** [comment]
If reading in random order is so slow, how can any sort algorithm improve that? Or is merge sort trying to do everything in chunks?

**Comment by 153957** [comment]
Well the problem here is reading chunks and using only a single row. If you could read a chunk and write all the rows in that chunk to the correct place on disk it would be more efficient, Im not certain if that can be enforced..

I believe what Tom is suggesting would be an step-by-step sorting? First sort multiple chunks and then sort them into the final file? like explained here https://en.wikipedia.org/wiki/External_sorting
That would means each chunk only needs to be read (and written) twice (or more often if more steps are required).. this might be faster, especially if the initial chunks to sort are chosen intelligently, but that would be another problem..

**Comment by 153957** [comment]
Possibly relevant; http://www.pytables.org/usersguide/optimization.html#id12
Random reading mean access time from a 15 GB data file, depending on chunksize and compression used.

Currently we are using no compression, and automatic chunksize.
Perhaps we should be using blosc and smaller chunksize for the intermediate HDF5 file (the not sorted one).

**Comment by davidfokkema** [comment]
Wow, external merge sort! That's nice. And using blosc and trying out smaller chunk sizes. Not a trivial thing to do, though.

**Comment by tomkooij** [comment]
I implemented a mergesort: https://github.com/tomkooij/lio-project/blob/6d5db84bc16203453df8758460fa8dd4209279ee/mergesort/mergesort.py

This just sorted a 30Gb corsika_unsorted.h5 in **100 minutes**, by:
- sequentially read the file in a large chunks 
- sort each part in memory using ```numpy.array.sort()``` 
- store each part in temporary table
- store an iterator that know's how to read this temporary table.
- merge all temp tables using ```heapq.merge()```

However, I did not store the results and I kept the tempfile continuously open. Python was using 33Gb of memory, so everything was still in cache.

I need to implement buffered write and test a file that is much larger than memory.

Creating the temporary tables can be parallellised using a ```multiprocessing.Pool()``` (but may be I/O limited with a single thread).
Optimising the size of the parts so that as much as possible is done in memory (and chosing the best algorithm) can improve as well.

I'm confident sorting of very large showers can be done within hours, not days.

**Comment by davidfokkema** [comment]
Wow, just wow. And so short and concise! How did you know about heapq.merge?

**Comment by tomkooij** [comment]
I just googled and stole the idea from csvsort:https://bitbucket.org/richardpenman/csvsort/
I need to add credit to this in the script.

**Comment by 153957** [comment]
Was that on an interactive stoomboot node?
You got to use 33 Gb of RAM (is that b or B?)?
Normal Stoomboot jobs only get 8 GB.
If I understand correctly, you were using that much RAM because you did not close/flush the TEMP file which caused most of it to remain in RAM?

**Comment by tomkooij** [comment]
Indeed, I deed no close (but did flush) the tempfile and it seems to remains in RAM. After every chunk is read the memory rises accordingly.

On an interactive node, it seems you can use 32 cores and 64Gb of real RAM. I'm not sure you are supposed to do that, but the node I use (stbc-i1) is largely unused, because it has been down/broken in the past.

**Comment by davidfokkema** [comment]
Ah, yes! It would be nice to add credit. You did some good googling, ;-)!

**Comment by tomkooij** [comment]
Memory usage was due to a bug: I passed the entire table.recarray to the iterator instead of just the table object... Memory usage is now under 500MBytes while using 350Mb temp tables. 

I implemented "quick-and-dirty" buffered write (instead of the slow row by row) and the 30 Gb testfile now sorts in about 2 hours on a single CPU in 500Mb memory.

**Comment by davidfokkema** [comment]
👏🎉

**Comment by 153957** [comment]
Could you make it into a PR for sapphire? as an alternative path in `sapphire.corsika.store_corsika_data.store_and_sort_corsika_data` for large files? (or always use mergesort if there is no great penalty for small files).

**Comment by tomkooij** [comment]
Will do. (I've got the branch almost finished locally. Should be able to finish it on/before monday.

**Comment by 153957** [comment]
Ok great! Well, worry about other things tomorrow 😉

**Comment by tomkooij** [comment]
Fixed in #118
